Issue #11166: Remove usage of getFileContents() from WriteTagCheck

Rohanraj123 opened the pull request:
part of #11166 

Diff Regression config: https://gist.githubusercontent.com/Rohanraj123/f8548a1a76258afebf6c9ef15d09dd59/raw/3d05b3263103cb1ab091134458ce50af7081f3c0/gistfile1.txt

Rohanraj123 commented:
GitHub, generate report for configs in PR description

Rohanraj123 commented:
Unexpected violation occurs at the empty lines. Looking into the problem, but please suggest if you find something related or some bug in code!!

Rohanraj123 commented:
Github, generate report

Rohanraj123 commented:
@romani PTAL

Rohanraj123 commented:
Hi @relentless-pursuit Sir, Can you please help reviewing this PR. It would help a lot. Thanks!

Rohanraj123 commented:
ping @romani

nrmancuso commented:
CI is the first reviewer, please make it happy first.

Zopsss commented:
> Unexpected violation occurs at the empty lines. 

Where violation is occurring?

Rohanraj123 commented:
Hey @nrmancuso @romani @Zopsss, Please take a look. Would address Pitest-warning in next PR.

nrmancuso commented on config/pitest-suppressions/pitest-utils-suppressions.xml:
We do not allow this list of suppressions to grow except in extreme cases. Please find a way to cover this.

Rohanraj123 replied:
Suppression was removed.

nrmancuso reviewed (changes requested):
Item:

Zopsss commented:
Cannot review PR until pitest suppression is removed. As Nick suggested, please find a way to fix it

Zopsss commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -186,6 +190,41 @@ public static String getJavadocCommentContent(DetailAST javadocCommentBegin) {
+
+                if (current.getType() == TokenTypes.BLOCK_COMMENT_BEGIN
+                        && JavadocUtil.isJavadocComment(current)) {
+                    result = current;
```
why not add `break` statement when result is found? No need to check for `result == null`

Rohanraj123 replied:
My bad, And this change will also resolve on PITEST warning. Now only one left. Will do it soon.

Rohanraj123 commented:
@nrmancuso @Zopsss Please take a look.

nrmancuso commented on config/pitest-suppressions/pitest-utils-suppressions.xml:
Why is this changed?

Rohanraj123 replied:
Will revert this change in the next push!

nrmancuso commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheck.java:
```diff
@@ -185,31 +184,42 @@ public int[] getAcceptableTokens() {
+
+        if (javadocComment != null) {
+            final String commentContent = JavadocUtil.getBlockCommentContent(javadocComment);
+            final Pattern lineSplitPattern = Pattern.compile("\\R");
```
Can we avoid compiling this pattern each time this method is called?

nrmancuso commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheck.java:
```diff
@@ -185,31 +184,42 @@ public int[] getAcceptableTokens() {
-     * @param comment the Javadoc comment for the type definition.
+     * @param javadocLineNo the starting line number of the Javadoc comment block.
+     * @param astLineNo the line number of the type definition
+     *                  (e.g., class, record, constructor).
```
A constructor is not a type definition.

Rohanraj123 replied:
my bad sorry ! removed the line.

nrmancuso commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -186,6 +190,42 @@ public static String getJavadocCommentContent(DetailAST javadocCommentBegin) {
+     *     or {@code null} if none is found.
+     */
+    @Nullable
+    public static DetailAST findJavadocComment(DetailAST ast) {
```
Few issues here:

1. Judging by the method name, I would expect this to return a `DetailNode`, not a `DetailAST`.
2. Why are we traversing the AST like this? We have Treewalker to handle traversal already.
3. We have no need to extend a util class until a method is needed in more than one place, if we actually need this method, it should be a private, static method in the check class.

Rohanraj123 replied:
> Why are we traversing the AST like this? We have Treewalker to handle traversal already.
Would look into that.

Rohanraj123 replied:
> Have you checked through existing utilities and checks for something like this (if we even need it)?

I removed this function, and use TreeWalker for AST traversal and using methods of BlockCommentPosition.java class for analysis of associated type def.

> We have no need to extend a util class until a method is needed in more than one place, if we actually need this method, it should be a private, static method in the check class.

Removed !

nrmancuso commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheck.java:
```diff
@@ -185,31 +184,42 @@ public int[] getAcceptableTokens() {
+            checkTag(javadocComment.getLineNo(), ast.getLineNo(), cmtLines);
         }
     }
 
     /**
-     * Verifies that a type definition has a required tag.
+     * Verifies that a type definition has a required Javadoc tag.
+     *
+     * <p>If the tag is found and correctly formatted,
+     * it logs at the line of the tag.
+     * If the tag is missing, it logs the violation
+     * at the line where the type definition begins.</p>
```
Please just update this to explain that this method validates the javadoc  comment against the given configuration. This check is strange to begin with, let's just avoid trying to reiterate it's logic here.

Rohanraj123 replied:
done !

nrmancuso commented on src/test/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtilTest.java:
```diff
@@ -114,6 +115,80 @@ public void testInlineTagMethodRef() {
     }
 
+    @Test
+    public void testFindJavadocComment() {
```
This is a red flag, we will not allow artificial tests like this except in very extreme cases (which this is not).

nrmancuso replied:
All of these tests will need to go away based on my above comment regardless.

Rohanraj123 replied:
done

nrmancuso reviewed (changes requested):
Items:

Zopsss commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -186,6 +190,42 @@ public static String getJavadocCommentContent(DetailAST javadocCommentBegin) {
+                    break;
+                }
+                else {
+                    DetailAST child = current.getLastChild();
```
Once Javadoc is found then use `getJavadocTags` to get all the tags

https://github.com/checkstyle/checkstyle/blob/7474b313dab18ec9daf5e65014ed599d7c819459/src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java#L96

and then process them.

Also as Nick suggested at https://github.com/checkstyle/checkstyle/pull/16757/files#r2041260952, we don't need util method if it isn't required in more than one place so move this logic to the Check itself.

Rohanraj123 replied:
Optimizing, how we process  child is okay. But should we made changes to the logic of checkTag?. Because if we use getJavadocTags(), then violations needed to handelled differently. Previously we were just using getJavadocBefore() method and I tried a replacement of that method with findJavadocComment().  Please confirm.

Rohanraj123 commented:
@Zopsss Please avoid small errors! Please just confirm the direction is correct , I tried to use TreeWalker for AST traversal.

Zopsss commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheck.java:
```diff
@@ -94,7 +101,7 @@
  *
  * @since 4.2
  */
-@StatelessCheck
+@FileStatefulCheck
```
@nrmancuso is it okay to convert the Check into Stateful check? I think the current approach is good just need your confirmation.

nrmancuso replied:
Best to avoid this when possible

Rohanraj123 replied:
done!

Rohanraj123 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheck.java:
```diff
@@ -182,34 +194,107 @@ public int[] getAcceptableTokens() {
+     *         {@code TokenTypes.CLASS_DEF}, etc.), or
+     *         {@code 0} if no associated definition is found.
+     */
+    private int associatedTypeDef(DetailAST ast) {
```
@Zopsss 
Here I think we have issue in our isOnAnnotationField().
Problem : 
Suppose In input file : 
1. There is a method ( class member ), `isOnAnnotationField` returns true as well as `isOnMethod` returns true.
2. If there is a annoation field ( functoin inside a annotation ) .  `isOnAnnotationField` returns true as well as `isOnMethod` returns true.

This behavious creates difficulty during mapping here in `associatedTypeDef()`. We need to make changes to isAnnotationField I believe.

Zopsss replied:
If you find a bug in some method, it is best to create a separate issue for it so we can discuss about it there, after that bug is resolved we can continue on this PR.

Rohanraj123 replied:
I would do that thanks!

Rohanraj123 commented:
> Always allow reviewers to resolve their comments; this allows us to make sure that each item is addressed
> Always reply to each and every comment, with either "done" or any questions/discussion you might have

I will keep in mind from next time !! Thanks for the guidance. 

> If you plan to continue to work on this PR, please do the above, and make CI happy. If not, please close it.

I will continue this PR after this issue I opened : https://github.com/checkstyle/checkstyle/issues/16900
Because it is creating problem in mapping to `method` and `AnnotationField`.

nrmancuso commented:
> I will keep in mind from next time !! Thanks for the guidance.

Please un-resolve each comment and reply to all

> I will continue this PR after this issue I opened : #16900 Because it is creating problem in mapping to `method` and `AnnotationField`.

I will assign this PR to @Zopsss to help you work through their request at https://github.com/checkstyle/checkstyle/pull/16757#discussion_r2050867495

Zopsss commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheck.java:
```diff
@@ -182,34 +194,107 @@ public int[] getAcceptableTokens() {
-            contents.getJavadocBefore(lineNo);
-        if (cmt != null) {
-            checkTag(lineNo, cmt.getText());
+        if (isJavadocComment(ast)) {
+            final int associatedType = associatedTypeDef(ast);
+
+            if (!isTokenConfigured(associatedType)) {
+                return;
+            }
```
your approach is not efficient IMO.

You're first checking if the AST is a javadoc or not, if it is javadoc then you're checking if it is for the configured token or not. Instead of this why not just directly check if the configured token has javadoc or not. If it has then proceed further. Try something like I mentioned at https://github.com/checkstyle/checkstyle/pull/16757#discussion_r2042642892

Zopsss commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheck.java:
```diff
@@ -166,6 +170,7 @@ public int[] getDefaultTokens() {
             TokenTypes.ENUM_DEF,
             TokenTypes.ANNOTATION_DEF,
             TokenTypes.RECORD_DEF,
+            TokenTypes.BLOCK_COMMENT_BEGIN,
```
We don't need BLOCK_COMMENT_BEGIN token anymore as we're manually checking if the configured token has Javadoc or not

Rohanraj123 replied:
Yes my bad!

Zopsss commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheck.java:
```diff
@@ -182,34 +187,85 @@ public int[] getAcceptableTokens() {
+        // Prefer Javadoc directly above the node
+        DetailAST cmt = ast.findFirstToken(TokenTypes.BLOCK_COMMENT_BEGIN);
+        if (cmt == null) {
+            // Check MODIFIERS and TYPE block for comments
+            final DetailAST modifiers = ast.findFirstToken(TokenTypes.MODIFIERS);
+            final DetailAST type = ast.findFirstToken(TokenTypes.TYPE);
+
+            cmt = modifiers.findFirstToken(TokenTypes.BLOCK_COMMENT_BEGIN);
+            if (cmt == null && type != null) {
+                cmt = type.findFirstToken(TokenTypes.BLOCK_COMMENT_BEGIN);
+            }
```
Why we need MODIFIERS & TYPE to find the block comment. 

https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#MODIFIERS

https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#TYPE

Rohanraj123 replied:
These are the places where javadoc can be found ( relevant to WriteTagCheck ). These are the types which can potentially contain javadoc. 
Javadoc can be : 
1. Direct children of ast.
2. Children of MODIFIERS.
3. Children of TYPE.

If required I can provide the Input files showcasing the edgecases.

Zopsss replied:
yes and also show their AST

Zopsss commented:
Please generate the diff to test the new logic @Rohanraj123

Rohanraj123 commented:
@nrmancuso @romani Please take a look. And Please let me know if anything else required .

Zopsss commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheck.java:
```diff
@@ -185,31 +189,82 @@ public int[] getAcceptableTokens() {
+    /**
+     * Validates the Javadoc comment against the configured requirements.
+     *
+     * @param javadocLineNo the starting line number of the Javadoc comment block.
+     * @param astLineNo the line number of the type definition.
+     * @param comment the lines of the Javadoc comment block.
```
First, `astLineNo` should come and then `javadocLineNo`.

Rohanraj123 replied:
my bad!

Zopsss replied:
This is not done yet

Rohanraj123 replied:
Now done . Sorry!

Zopsss commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java:
```diff
@@ -3,7 +3,7 @@
 tagSeverity = error
 tagFormat = \\S
-tokens = INTERFACE_DEF, CLASS_DEF, ENUM_DEF, ANNOTATION_DEF, RECORD_DEF, COMPACT_CTOR_DEF, CTOR_DEF
+tokens = CLASS_DEF, RECORD_DEF, COMPACT_CTOR_DEF, CTOR_DEF
```
Why this was changed?

Rohanraj123 replied:
Because they were not needed for the test cases. Will revert if not required.

Zopsss commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTag.java:
```diff
@@ -3,7 +3,7 @@
 tagFormat = \\S
 tagSeverity = error
-tokens = (default)INTERFACE_DEF, CLASS_DEF, ENUM_DEF, ANNOTATION_DEF, RECORD_DEF
+tokens = INTERFACE_DEF, CLASS_DEF, ENUM_DEF, ANNOTATION_DEF, RECORD_DEF
```
Why this was changed?

Rohanraj123 replied:
Reverted!

Zopsss reviewed (changes requested):
Sorry for the delay, was busy with some personal work. 

Last minors:

Zopsss commented:
@Rohanraj123 ping

Rohanraj123 commented:
> @Rohanraj123 ping

Anything else am I supposed to do here? @Zopsss. I addressed all your review pointers. 
@romani @nrmancuso Please have a look .

Zopsss commented:
> Anything else am I supposed to do here?

The changes I requested are not done yet and I have also pinged you in one comment, please reply there too

Zopsss reviewed (approved):
PR is good from my side. @nrmancuso please review it.

Zopsss commented:
@nrmancuso ping

Rohanraj123 commented:
@nrmancuso ping

romani commented:
@Rohanraj123 , please do https://github.com/checkstyle/checkstyle/pull/16757/files#r2100484288
And we will merge

Rohanraj123 commented:
> @Rohanraj123 , please do https://github.com/checkstyle/checkstyle/pull/16757/files#r2100484288 And we will merge

Done!

romani reviewed (approved):
Good step forward.

It will help in future bigger redesign of this Check
